File: src/main/java/ar/com/florius/aao/SafeTag.java
package ar.com.florius.aao;

import ar.com.florius.aao.semilattice.TagName;

import java.util.List;

public class SafeTag<T> implements Taggable<T> {
    private final T value;
    private final TagName tag;

    public SafeTag(T value, TagName tag) {
        this.value = value;
        this.tag = tag;
    }

    public static <T> T untag(SafeTag<T> tagged) {
        return tagged.getValue();
    }

    public static <T> TagName getTag(SafeTag<T> tagged) {
        return tagged.getTag();
    }

    @Override
    public TagName getTag() {
        return tag;
    }

    @Override
    public T getValue() {
        return value;
    }

    public TagName operate(List<TagName> argsTag) {
        TagName newTag = argsTag.stream().reduce(this.getTag(), TagName::join);
        if (isIncompatible(newTag)) {
            throw new IncompatibleTagsException("Tags are incompatible between this (" + this.getTag() + ") and the arguments (" + argsTag + ")");
        }
        return newTag;
    }

    private boolean isIncompatible(TagName newTag) {
        return newTag.equals(TagName.Incompatible.INSTANCE);
    }
}

File: src/main/kotlin/ar/com/florius/aao/SafeTag.kt
package ar.com.florius.aao

import ar.com.florius.aao.semilattice.Namespace
import java.lang.reflect.Method

class SafeTag<T>(override val value: T, override val tag: Namespace) : Taggable<T, Namespace> {
    fun <A, RETURN> apply(f: (t: T, op1: A) -> RETURN, op1: SafeTag<A>): SafeTag<RETURN> {
        val newTag = tag.join(op1.tag)
        return SafeTag(f(this.value, op1.value), newTag)
    }

    fun <A, B, RETURN> apply(f: (t: T, op1: A, op2: B) -> RETURN, op1: SafeTag<A>, op2: SafeTag<B>): SafeTag<RETURN> {
        val newTag = tag.join(listOf(op1.tag, op2.tag))
        return SafeTag(f(this.value, op1.value, op2.value), newTag)
    }

    fun apply(method: Method, args: List<Taggable<Any?, Namespace>>): SafeTag<Any?> {
        val newTag = tag.join(args.map { it.tag })
        val result = method.invoke(this.value, *args.map { it.value }.toTypedArray())
        return SafeTag(result, newTag)
    }

    override fun toString(): String {
        return "$value«$tag»"
    }


    companion object {
        @JvmStatic
        fun <T> untag(tagged: SafeTag<T>): T {
            return tagged.value
        }

        @JvmStatic
        fun <T> getTag(tagged: SafeTag<T>): Namespace {
            return tagged.tag
        }
    }
}